Signal/src/contact/OWSContactsManager.m

```diff
@@ -206,30 +160,148 @@ - (void)intersectContactsWithRetryDelay:(double)retryDelaySeconds
                                                                            failure:failure];
 }
 
-- (void)fireSignalRecipientsDidChange
+- (void)pullLatestAddressBook {
+    dispatch_async(ADDRESSBOOK_QUEUE, ^{
+        CFErrorRef creationError = nil;
+        ABAddressBookRef addressBookRef = ABAddressBookCreateWithOptions(NULL, &creationError);
+        checkOperationDescribe(nil == creationError, [((__bridge NSError *)creationError)localizedDescription]);
+        ABAddressBookRequestAccessWithCompletion(addressBookRef, ^(bool granted, CFErrorRef error) {
+            if (!granted) {
+                [OWSContactsManager blockingContactDialog];
+            }
+        });
+        NSArray<Contact *> *contacts = [self getContactsFromAddressBook:addressBookRef];
+        [self updateWithContacts:contacts];
+    });
+}
+
+- (void)updateWithContacts:(NSArray<Contact *> *)contacts
 {
-    AssertIsOnMainThread();
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
 
-    [[NSNotificationCenter defaultCenter] postNotificationName:OWSContactsManagerSignalRecipientsDidChangeNotification
-                                                        object:nil];
+        NSMutableDictionary<NSString *, Contact *> *allContactsMap = [NSMutableDictionary new];
+        for (Contact *contact in contacts) {
+            for (PhoneNumber *phoneNumber in contact.parsedPhoneNumbers) {
+                NSString *phoneNumberE164 = phoneNumber.toE164;
+                if (phoneNumberE164.length > 0) {
+                    allContactsMap[phoneNumberE164] = contact;
+                }
+            }
+        }
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            self.allContacts = contacts;
+            self.allContactsMap = [allContactsMap copy];
+
+            [self intersectContacts];
+
+            [self updateSignalAccounts];
```

Comment:
Yes, but:

a) the contacts intersection is asynchronous and we'd like to update the views immediately.
b) the contacts intersection may fail.

Updating the views should be cheap enough that it's okay to refresh them twice whenever there's a contacts change, especially since contacts changes should be rare.